Variable size in EC2 API request and ifdefs for linux build

ZackaryWelch opened the pull request.

ZackaryWelch committed:
Added .gitignore

ZackaryWelch committed:
Snapshot size now determined by image file size. Added ifdefs for linux build.

ZackaryWelch commented:
https://github.com/cperciva/bsdec2-image-upload/issues/9

cperciva commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 *.d
 *.o
 bsdec2-image-upload
+.depend
```
This looks like it belonged in the previous commit.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -1,6 +1,10 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
+#ifdef __linux__
```
For portability, make this (and the call which uses it later) `#ifndef SO_NOSIGPIPE`.

cperciva replied:
Still need to fix this.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /* Linux does not support SO_NOSIGPIPE
```
```
/*
 * Multi-line comments should
 * be formatted like this.
 */
```

cperciva replied:
Also, sentences end with `.`.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
-		return "Could not disable SIGPIPE";
+    /* Linux does not support SO_NOSIGPIPE
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally */
+	#ifdef __linux__
```
`#ifndef SO_NOSIGPIPE`, and also preprocessor directives should be in column 1.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
+    /* Linux does not support SO_NOSIGPIPE
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally */
+	#ifdef __linux__
+		signal(SIGPIPE, SIG_IGN);
```
Error handling?

cperciva commented on main.c:
```diff
@@ -21,8 +21,13 @@
 #include "warnp.h"
 
 #ifndef CERTFILE
+/* Different cert locations between Linux and BSD */
```
This should be handled by defining CERTFILE in the `make` invocation.

cperciva commented on main.c:
```diff
@@ -1256,12 +1261,22 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
-	}
+
+    /* Linux does not support MAP_NOCORE */
+    #ifdef __linux__
```
I would change this to
```
#ifndef MAP_NOCORE
#define MAP_NOCORE 0
#endif
```

cperciva commented on main.c:
```diff
@@ -1334,7 +1349,7 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
 registerimage(const char * region, const char * snapshot, const char * name,
     const char * desc, const char * arch, int sriov, int ena,
-    const char * key_id, const char * key_secret)
+    const char * key_id, const char * key_secret, const char * diskimage)
```
We're getting the file size in uploadvolume and uploadsnap.  Return the size from those functions and pass it into this one rather than statting the file a second time.

ZackaryWelch replied:
How should this be done?

cperciva replied:
Add a "imgsz" parameter to each of the functions?

ZackaryWelch committed:
Updates per review

ZackaryWelch committed:
Added conditional build for Linux

cperciva commented on Makefile:
```diff
@@ -2,10 +2,18 @@ PROG=	bsdec2-image-upload
+UNAME_S := $(shell uname -s)
+.if (${UNAME_S},Linux)
+CERTFILE = /etc/ssl/certs/ca-bundle.crt
+CFLAGS += --no-warnings
```
Umm... no.

ZackaryWelch replied:
How would you suggest defining CERTFILE in the makefile?

cperciva commented on Makefile:
```diff
@@ -40,5 +48,6 @@ IDIRS	+=	-I lib/util
 
 CFLAGS	+=	-g
 CFLAGS	+=	${IDIRS}
+CFLAGS  += -DCERTFILE
```
What do you think this does, exactly?

ZackaryWelch replied:
https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html

ZackaryWelch replied:
Oh lol, I'm missing the assignment

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally.
+     */
+#ifndef SO_NOSIGPIPE
+		if (signal(SIGPIPE, SIG_IGN) == SIG_ERR)
```
Superfluous indentation.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
-		return "Could not disable SIGPIPE";
+    /*
+     * Linux does not support SO_NOSIGPIPE.
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally.
```
Please rewrap.  There's no need for each sentence to be on its own line.

cperciva commented on main.c:
```diff
@@ -1256,12 +1258,12 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
-		warnp("Could not map disk image");
-		goto err1;
-	}
+    /* Map the file. */
```
Why mangle the indentation?

ZackaryWelch committed:
Updated per revisions

Fixed indentations and replaced spaces with tabs in Makefile
Assigned CERTFILE in Makefile
Changed __linux__ to #ifndef SO_NOSIGPIPE
Removed extra file stat. File size is now passed between functions.

cperciva commented on Makefile:
```diff
@@ -2,10 +2,18 @@ PROG=	bsdec2-image-upload
+UNAME_S	:=	$(shell uname -s)
+.if (${UNAME_S},Linux)
+CERTFILE	=	\"/etc/ssl/certs/ca-bundle.crt\"
+CFLAGS	+=	--no-warnings
```
Still not sure why you're disabling all compiler warnings on Linux.

ZackaryWelch replied:
What does -Wno-error=\\#warnings do? That was my best guess as to what it did.

cperciva commented on lib/util/sslreq.c:
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /*
```
Please use tabs for indentation, like the rest of the file does.

cperciva commented on main.c:
```diff
@@ -1255,10 +1259,11 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
 	/* Map the file. */
 	if ((p = mmap(NULL, len, PROT_READ, MAP_PRIVATE | MAP_NOCORE,
-	    fileno(f), 0)) == MAP_FAILED) {
+		fileno(f), 0)) == MAP_FAILED) {
```
Why did you change the indentation here?

ZackaryWelch replied:
I have no idea how this happened

cperciva commented on main.c:
```diff
@@ -1350,6 +1355,10 @@ registerimage(const char * region, const char * snapshot, const char * name,
 		goto err2;
+    
+    /* Get image size in GB and set to 10 if less than 10. */
+    size_t gb_size = (*imgsz)/(1024 * 1024 * 1024);
```
Put variable declarations at the top of the function.

Also, the indentation is wrong.

Also, there should be whitespace on either side of the `/` character.

Also, `if(gb_size < 10) gb_size = 10;` should be
```
if (gb_size < 10)
        gb_size = 10;
````

cperciva commented on main.c:
```diff
@@ -1364,7 +1373,7 @@ registerimage(const char * region, const char * snapshot, const char * name,
 	    "BlockDeviceMapping.1.Ebs.SnapshotId=%s&"
 	    "BlockDeviceMapping.1.Ebs.VolumeType=gp2&"
-	    "BlockDeviceMapping.1.Ebs.VolumeSize=10&"
+	    "BlockDeviceMapping.1.Ebs.VolumeSize=%d&"
```
`%d` is the wrong type for printing a `size_t` value.

cperciva commented on main.c:
```diff
@@ -2102,14 +2111,16 @@ main(int argc, char * argv[])
 		nregions = 1;
 	}
 
+	size_t *imgsz = malloc(sizeof(size_t));
```
Umm, what?  Just have `size_t imgsz` and pass `&imgsz` to functions below.

ZackaryWelch commented:
Awesome! This all appears implemented now.
